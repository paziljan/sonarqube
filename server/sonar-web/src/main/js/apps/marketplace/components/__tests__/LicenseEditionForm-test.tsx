/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { click } from '../../../../helpers/testUtils';
import LicenseEditionForm from '../LicenseEditionForm';

jest.mock('../../../../api/marketplace', () => ({
  applyLicense: jest.fn(() =>
    Promise.resolve({ nextEditionKey: 'foo', installationStatus: 'AUTOMATIC_IN_PROGRESS' })
  )
}));

const applyLicense = require('../../../../api/marketplace').applyLicense as jest.Mock<any>;

const DEFAULT_EDITION = {
  key: 'foo',
  name: 'Foo',
  desc: 'Foo desc',
  download_link: 'download_url',
  more_link: 'more_url',
  request_license_link: 'license_url'
};

beforeEach(() => {
  applyLicense.mockClear();
});

it('should display correctly', () => {
  expect(getWrapper()).toMatchSnapshot();
});

it('should correctly change the button based on the status', () => {
  const wrapper = getWrapper();
  (wrapper.instance() as LicenseEditionForm).mounted = true;
  wrapper.setState({ status: 'NO_INSTALL' });
  expect(wrapper.find('button')).toMatchSnapshot();
  wrapper.setState({ status: 'AUTOMATIC_INSTALL' });
  expect(wrapper.find('button')).toMatchSnapshot();
  wrapper.setState({ status: 'MANUAL_INSTALL' });
  expect(wrapper.find('button')).toMatchSnapshot();
});

it('should update the edition status after install', async () => {
  const updateEditionStatus = jest.fn();
  const wrapper = getWrapper({ updateEditionStatus });
  const form = wrapper.instance() as LicenseEditionForm;
  form.mounted = true;
  form.handleLicenseChange('mylicense', 'AUTOMATIC_INSTALL');
  click(wrapper.find('button'));
  expect(applyLicense).toHaveBeenCalledWith({ license: 'mylicense' });
  await new Promise(setImmediate);
  expect(updateEditionStatus).toHaveBeenCalledWith({
    nextEditionKey: 'foo',
    installationStatus: 'AUTOMATIC_IN_PROGRESS'
  });
});

function getWrapper(props = {}) {
  return shallow(
    <LicenseEditionForm
      edition={DEFAULT_EDITION}
      editions={[DEFAULT_EDITION]}
      onClose={jest.fn()}
      updateEditionStatus={jest.fn()}
      {...props}
    />
  );
}
