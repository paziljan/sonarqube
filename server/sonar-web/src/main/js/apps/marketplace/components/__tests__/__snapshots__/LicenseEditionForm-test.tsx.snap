// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly change the button based on the status 1`] = `
<button
  className="js-confirm"
  disabled={false}
  onClick={[Function]}
>
  save
</button>
`;

exports[`should correctly change the button based on the status 2`] = `
<button
  className="js-confirm"
  disabled={false}
  onClick={[Function]}
>
  marketplace.install
</button>
`;

exports[`should correctly change the button based on the status 3`] = `
<button
  className="js-confirm"
  disabled={false}
  onClick={[Function]}
>
  save
</button>
`;

exports[`should display correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="marketplace.install_x.Foo"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      marketplace.install_x.Foo
    </h2>
  </header>
  <LicenseEditionSet
    className="modal-body"
    edition={
      Object {
        "downloadUrl": "download_url",
        "homeUrl": "more_url",
        "key": "foo",
        "name": "Foo",
        "requestUrl": "license_url",
        "textDescription": "Foo desc",
      }
    }
    editions={
      Array [
        Object {
          "downloadUrl": "download_url",
          "homeUrl": "more_url",
          "key": "foo",
          "name": "Foo",
          "requestUrl": "license_url",
          "textDescription": "Foo desc",
        },
      ]
    }
    updateLicense={[Function]}
  />
  <footer
    className="modal-foot"
  >
    <a
      className="js-modal-close"
      href="#"
      onClick={[Function]}
    >
      cancel
    </a>
  </footer>
</Modal>
`;
